File handling refactoring

jbuening opened the pull request:
Just a few changes to remove duplicated code, changing the signature of compressed_fd_ostream and fixing two error messages to reflect the actual file extension.

251 commented on lib/Support/CompressionStream.cpp:
```diff
@@ -23,17 +23,17 @@
 namespace klee {
 
-compressed_fd_ostream::compressed_fd_ostream(const char *Filename,
+compressed_fd_ostream::compressed_fd_ostream(std::string &Filename,
```
Why not const &?

251 commented on lib/Core/Executor.cpp:
```diff
@@ -362,33 +363,20 @@ Executor::Executor(LLVMContext &ctx, const InterpreterOptions &opts,
-    debugInstFile =
-        new llvm::raw_fd_ostream(debug_file_name.c_str(), ErrorInfo);
-#endif
+      debugInstFile = klee_open_output_file(debug_file_name, error);
```
unique_ptr for debugInstFile instead of raw pointer

jbuening replied:
done

251 commented on lib/Solver/QueryLoggingSolver.cpp:
```diff
@@ -39,27 +40,18 @@ QueryLoggingSolver::QueryLoggingSolver(Solver *_solver, std::string path,
-#else
-    os = new llvm::raw_fd_ostream(path.c_str(), ErrorInfo);
-#endif
+    os = klee_open_output_file(path, error);
```
unique_ptr for os

jbuening replied:
done

251 commented on lib/Core/Executor.cpp:
```diff
@@ -362,33 +363,20 @@ Executor::Executor(LLVMContext &ctx, const InterpreterOptions &opts,
     } else {
-      debugInstFile = new compressed_fd_ostream(
-          (debug_file_name + ".gz").c_str(), ErrorInfo);
+      debug_file_name = debug_file_name + ".gz";
```
`+=` or just use `.append()`

251 commented on lib/Solver/QueryLoggingSolver.cpp:
```diff
@@ -39,27 +40,18 @@ QueryLoggingSolver::QueryLoggingSolver(Solver *_solver, std::string path,
 #ifdef HAVE_ZLIB_H
   } else {
-    os = new compressed_fd_ostream((path + ".gz").c_str(), ErrorInfo);
+    path = path + ".gz";
```
see comment above

251 commented on lib/Support/CompressionStream.cpp:
```diff
@@ -23,17 +23,17 @@
 #if LLVM_VERSION_CODE >= LLVM_VERSION(3, 5)
   std::error_code EC =
-      llvm::sys::fs::openFileForWrite(Filename, FD, llvm::sys::fs::F_None);
+      llvm::sys::fs::openFileForWrite(Filename.c_str(), FD, llvm::sys::fs::F_None);
```
Did you try just `Filename` - at least in newer versions there is a Twine ctor for std::string

jbuening commented:
@251 Thanks for your comments. Ok, let's actually do some real refactoring in this PR then. I started addressing your comments, but have found some related things I plan to look into. Will push a new version of this later tomorrow.

jbuening commented:
@251: I just pushed a new patch, addressing your comments, refactoring `klee_open_output_file()` to return a `unique_ptr` and everything that follows from that. I also introduced a function called `klee_open_compressed_file()` to deal with compressed files in the same way and fixed some small issues along the way.

jbuening commented on lib/Core/StatsTracker.cpp:
```diff
@@ -243,12 +241,13 @@ StatsTracker::StatsTracker(Executor &_executor, std::string _objectFilename,
-      executor.addTimer(new WriteStatsTimer(this), StatsWriteInterval);
+      if (StatsWriteInterval > 0)
+        executor.addTimer(new WriteStatsTimer(this), StatsWriteInterval);
+    }
```
The logic here is that you need to set it to 0 to disable it (in case you want to use `-stats-write-after-instructions`), as otherwise the following issues an error message:
https://github.com/klee/klee/blob/224532d66198c80c89b6017e14de4bbed53e305e/lib/Core/StatsTracker.cpp#L187-L190
Values other than 0 are not possible (in the else branch), as `StatsWriteInterval` is `unsigned`.

251 replied:
That's not what I meant (also this will probably change). I meant you should handle the error case for `if (statsFile)` and `if (istatsFile)`.

jbuening replied:
Oh, I see. Sorry, will change that.

jbuening commented on include/klee/Internal/Support/FileHandling.h:
```diff
@@ -6,14 +6,22 @@
-#define __KLEE_FILE_HANDLING_H__
+
+#ifndef INCLUDE_KLEE_INTERNAL_SUPPORT_FILEHANDLING_H_
+#define INCLUDE_KLEE_INTERNAL_SUPPORT_FILEHANDLING_H_
```
I certainly did not look hard enough :D I changed it to `KLEE_FILEHANDLING_H` and `KLEE_COMPRESSIONSTREAM_H` as most files seem to use this pattern.

jbuening commented on lib/Support/FileHandling.cpp:
```diff
@@ -7,33 +7,51 @@
-llvm::raw_fd_ostream *klee_open_output_file(std::string &path,
-                                            std::string &error) {
-  llvm::raw_fd_ostream *f;
+std::unique_ptr<llvm::raw_fd_ostream>
```
I changed it, now we just have to pretend it actually says `make_unique` ;-)

jbuening commented on tools/klee/main.cpp:
```diff
@@ -372,12 +371,12 @@ std::string KleeHandler::getOutputFilename(const std::string &filename) {
+  if (!f) {
     klee_warning("error opening file \"%s\".  KLEE may have run out of file "
                  "descriptors: try to increase the maximum number of open file "
                  "descriptors by using ulimit (%s).",
```
done

jbuening commented:
@251 Thanks for reviewing my changes. I should have addressed all your comments, except for the ones about the error cases in `StatsTracker.cpp` (see above).

jbuening commented:
@251 I added the two error messages now, should have looked at the full diff, not just the preview.

jbuening commented on tools/klee/main.cpp:
```diff
@@ -444,88 +443,85 @@ void KleeHandler::processTestCase(const ExecutionState &state,
+      if (f) {
+        for (auto &entry : cov) {
+          for (auto &line : entry.second) {
+            *f << *entry.first << ":" << line << '\n';
```
fixed!

251 commented:
Thanks, looks good to me then.

ccadar commented:
Nice patch, thanks, @jbuening.  And thanks, @251 for reviewing it.  Let's merge it now.

ccadar commented:
(I'll actually merge it tomorrow in case there are any other comments, but I think this is ready to go)

251 commented on tools/klee/main.cpp:
```diff
@@ -444,88 +443,85 @@ void KleeHandler::processTestCase(const ExecutionState &state,
-        *f << *I << "\n";
+      auto f = openTestFile("path", id);
+      if (f) {
+        for (auto &branch : concreteBranches) {
```
Actually, this range based for loop and the ones below might work with `const auto &`. Just nitpicking... ;)

jbuening replied:
Sure, I changed it

ccadar commented:
OK, time to merge!  Thanks again, @jbuening.
